modules/meshrenderinggl/glsl/oit/sort.glsl

```diff
@@ -105,27 +105,33 @@ void fillFragmentArraySortedUntilDepth(uint pixelIdx, float maxDepth, out int nu
     // Load fragments into a local memory array for sorting
     numFrag = 0;
     float depth = 0.0;
+    uint lastPtr = 0;
 
     while (numFrag < ABUFFER_SIZE) {
-        vec4 val = selectionSortNext(pixelIdx, depth);
+        vec4 val = selectionSortNext(pixelIdx, depth, lastPtr);
         depth = val.y;
         if (depth < 0.0 || depth > maxDepth) return;
         fragmentList[numFrag] = val;
         numFrag++;
     }
 }
 
-vec4 selectionSortNext(uint headPtr, float lastDepth) {
+vec4 selectionSortNext(uint headPtr, float lastDepth, inout uint lastPtr) {
     vec4 closestVal = vec4(-1.0);
     float minDepth = 1.0;
+    bool passedLastPtr = false;
+    uint minPtr = 0;
     while (headPtr != 0) {
         vec4 val = readPixelStorage(headPtr - 1);
-        if (val.y > lastDepth && val.y < minDepth) {
+        if (headPtr == lastPtr) passedLastPtr = true;
+        else if (val.y < minDepth && (val.y > lastDepth || (passedLastPtr && val.y == lastDepth) )) {
```

Comment:
This is a bit hard to understand. Consider adding comments about the reasoning here.

Reply:
Comments added in a345371f90a65a87225604c36ef357e9879e292e.
Basically, this all is here to solve the problem of fragments at the exact same depth.
Instead of relying on the depth for order alone, we need to keep tabs on what the last selected fragment was. In case we encounter the same depth (as `lastDepth`), we need to make sure it is not the same fragment (therefor, taking and updating `lastPtr`). Also, we need to make sure that fragment comes behind the last one, in order of list traversal - else, we'd get an endless loop between any two fragments of same depth. Therefor we save `passedLastPtr`.